crates/subgraph/src/apy.rs

```diff
@@ -0,0 +1,760 @@
+            // return early as soon as a match is found
+            return Ok(order_apy);
+        }
+    }
+
+    Ok(order_apy)
+}
+
+/// Calculates each token vault apy at the given timeframe
+/// Trades must be sorted indesc order by timestamp, this is
+/// the case if queried from subgraph using this lib functionalities
+pub fn get_token_vaults_apy(
+    trades: &[Trade],
+    vols: &[VaultVolume],
+    start_timestamp: Option<u64>,
+    end_timestamp: Option<u64>,
+) -> Result<Vec<TokenVaultAPY>, OrderbookSubgraphClientError> {
+    let mut token_vaults_apy: Vec<TokenVaultAPY> = vec![];
+    for vol in vols {
+        // this token vault trades in desc order by timestamp
+        let vault_trades = trades
+            .iter()
+            .filter(|v| {
+                (v.input_vault_balance_change.vault.vault_id.0 == vol.id
+                    && v.input_vault_balance_change.vault.token == vol.token)
+                    || (v.output_vault_balance_change.vault.vault_id.0 == vol.id
+                        && v.output_vault_balance_change.vault.token == vol.token)
+            })
+            .collect::<Vec<&Trade>>();
+
+        // this token vault first trade, indictaes the start time
+        // to find the end of the first day to find the starting capital
+        let first_trade = vault_trades[vault_trades.len() - 1];
+        let first_day_last_trade = vault_trades
+            .iter()
+            .filter(|v| {
+                u64::from_str(&v.timestamp.0).unwrap()
+                    <= u64::from_str(&first_trade.timestamp.0).unwrap()
+                        + TimeDelta::days(1).num_seconds() as u64
+            })
+            .collect::<Vec<&&Trade>>()[0];
+
+        // vaults starting capital at end of first day of its first ever trade
+        // as 18 point decimals
+        let vault_balance_change = if first_day_last_trade
+            .input_vault_balance_change
+            .vault
+            .vault_id
+            .0
+            == vol.id
+            && first_day_last_trade.input_vault_balance_change.vault.token == vol.token
+        {
+            &first_day_last_trade.input_vault_balance_change
+        } else {
+            &first_day_last_trade.output_vault_balance_change
+        };
+        let starting_capital = U256::from_str(&vault_balance_change.new_vault_balance.0)
+            .ok()
+            .and_then(|amount| {
+                to_18_decimals(
+                    ParseUnits::U256(amount),
+                    vault_balance_change
+                        .vault
+                        .token
+                        .decimals
+                        .as_ref()
+                        .map(|v| v.0.as_str())
+                        .unwrap_or("18"),
+                )
+                .ok()
+            });
+
+        // convert net vol to 18 decimals point
+        let net_vol = to_18_decimals(
+            ParseUnits::I256(vol.net_vol),
+            vol.token
```

Comment:
this still doesn't look like it has been moved to an impl

to be clear, i'm expecting the implementation of how decimals fallback and are parsed to be handled in one place - an impl